### ISSUE TITLE
[css-animations] Serialize all keyframe names as strings

### ISSUE THREAD

**Issue by csnardi** [w3c/csswg-drafts#2435]
After #118 and #801, support for naming a keyframe the same as a reserved keyword by using a `<string>` was added to parsing and CSSOM setting. [The WPT test](https://github.com/w3c/web-platform-tests/blob/master/css/cssom/CSSKeyframesRule.html#L59-L69) for this part of the spec seems to assume that if a keyword could be parsed as an `<ident>`, then it would be serialized as a string (with quotation marks); otherwise, no quotation marks would be used. However, this seems needlessly complex; wouldn't it be easiest just to serialize all keyframe names as strings? That way, it would be ensured that re-parsing the output would generate an identical rule, without having to check if the name is an `<ident>`.

**Comment by fantasai** [comment]
This is likely Web-incompatible, since Firefox and Blink currently return bare identifiers in getComputedStyle.
It's also inconsistent with the principle that CSS-internal user-chosen names use identifiers and strings are only used when interfacing with an external system.

I think what would make more sense is to exclude the CSS-wide keywords and 'none' from being animation names, as the `<custom-ident>` type recommends. We could still parse in strings as required by https://github.com/w3c/csswg-drafts/issues/118 but they would compute to `<custom-ident>` and any that match an excluded keyword would be invalid. This is effectively what we do for line names in CSS Grid, for example.

**Comment by ewilligers** [comment]
Do we need a more precise resolution before updating the spec and tests? I gather from a [PR discussion](https://github.com/w3c/csswg-drafts/pull/3296) that we need to serialize as strings whenever we do not have a valid animation-name custom identifier (e.g. `'"Initial"'`, `'"None"'`, `'" x "'`). Also, do we want to ban the empty string `'""'` ?

**Comment by yisibl** [comment]
Is there any progress here?

**Comment by javifernandez** [comment]
Is the CSSWG resolution mentioned https://github.com/w3c/csswg-drafts/issues/2435#issuecomment-428646888 still valid ? 

I've seen that WebKit has [updated the implementation](https://github.com/WebKit/WebKit/pull/38484) to match the current spec, which is what Firefox already implements; Chrome diverges, though.

**Comment by javifernandez** [comment]
@fantasai is the PR #12985 enough to cover the requested edits ? Id so, perhaps we can close this issue.

**Comment by cdoublev** [comment]
I am confused. :worried:

Quoting the new text:

  > The `<string>` should be treated as `<custom-ident>`, hence parsed as invalid if is one of the CSS-wide keywords or "default"

This is not what I understand from the above resolution, tests on WPT, and output in Firefox.

`"<css-wide-keyword>"`, `"default"`, `"none"`, `"1"`, `"not a custom ident"`, etc... are all valid on WPT and in Firefox, which serializes them as an identifier when possible. `normal` and `auto` are valid. `none` is invalid.

(edit) `""` (empty `<string>`) is also invalid in Firefox, which is not tested on WPT and was not specified before.

**Comment by javifernandez** [comment]
> This is not what I understand from the above resolution, tests on WPT, and output in Firefox.

I the discussion log this was the proposal made before the resolution:

> fantasai: Proposal is everywhere issue #﻿118 says we can accept a string we treat it at parse time as an ident

And then in the discussion held in https://github.com/w3c/csswg-drafts/pull/3296#issuecomment-437718506 it was mentioned again: 

> if ```<string>``` is one of the CSS-wide keywords or "default" then it is invalid (at parse / variable-resolution time) as per the rules for custom-idents: https://www.w3.org/TR/css-values-3/#custom-idents

It's true that the WPT have not been updated yet, but I think the spec should be updated first.

**Comment by cdoublev** [comment]
> It's true that the WPT have not been updated yet, but I think the spec should be updated first.

I assume that Webkit and Firefox should change their implementation, and there is no back-compat issue.

If `<string>` value must match `<custom-ident>`, when does it need to serialize as `<string>`?

**Comment by javifernandez** [comment]
> > It's true that the WPT have not been updated yet, but I think the spec should be updated first.
> 
> I assume that Webkit and Firefox should change their implementation, and there is no back-compat issue.

Yes, it was my understanding that they agreed on the CSSWG resolution and this interpretation. Regarding the web compat issue, it's been mentioned during the discussion as well:

> myles_: Web compat risk?
> Rossen_> 'auto'
> fantasai: If someone relies o n web animation name as 'initial' 'inherit' or 'none' yes. Otherwise no.

I don't think using a css-wide keyword as an animation identifier is a common use case.

> 
> If `<string>` value must match `<custom-ident>`, when does it need to serialize as `<string>`?

As far as I have understood, only quoted strings (eg, "a\ custom\ ident")

**Comment by nt1m** [comment]
That is not how I interpreted the resolution fwiw. Nothing in the resolution says that `<string>` should get custom-ident restrictions. The resolution only says to serialize to strings to the extent possible?

It's not a hard change to make either way, but it's rather a waste of time and unnecessary compat risk to change 2 implementations as opposed to one. It wouldn't be unreasonable for a website to name their animation "default" fwiw.

**Comment by javifernandez** [comment]
>  It wouldn't be unreasonable for a website to name their animation "default" fwiw.

I think the resolution tried to make invalid the use of "inherit" as valid animation-name, or "unset". Given that were invalid as custom-ident, it'd made sense to consider those strings as invalid. 

Additionally, if those strings are going to be serialized as identifiers, wouldn't be confusing to see the computed value as "inherit" ? just the subtle different of the use of "" to differentiate the reserved keyword. 

In any case, I think it's understandable to match what 2 implementors already have, but at least we should be able to get a new agreement now.

**Comment by javifernandez** [comment]
> That is not how I interpreted the resolution fwiw. Nothing in the resolution says that `<string>` should get custom-ident restrictions. The resolution only says to serialize to strings to the extent possible?

You meant "serialize to identifiers, to the extent possible" right ?

**Comment by nt1m** [comment]
> > That is not how I interpreted the resolution fwiw. Nothing in the resolution says that `<string>` should get custom-ident restrictions. The resolution only says to serialize to strings to the extent possible?
> 
> You meant "serialize to identifiers, to the extent possible" right ?

Yes. If it's not possible, it should serialize as a string

**Comment by javifernandez** [comment]
BTW, it seems that both Firefox and Safari accept CSS-wide keywords as valid, serialized as identifiers, although the computed value is 'none'. This is not spec compliant, right ?

**Comment by nt1m** [comment]
> BTW, it seems that both Firefox and Safari accept CSS-wide keywords as valid, serialized as identifiers, although the computed value is 'none'. This is not spec compliant, right ?

That sounds like a bug

**Comment by javifernandez** [comment]
> > BTW, it seems that both Firefox and Safari accept CSS-wide keywords as valid, serialized as identifiers, although the computed value is 'none'. This is not spec compliant, right ?
> 
> That sounds like a bug

Nah, nevermid, css-wide keywords are always accepted, although ignored if they don't apply to the specific property. 
I think it's all clear now. Thanks !

**Comment by yisibl** [comment]
@javifernandez So all of these are valid, right?

```
animation-name: "default";
animation-name: "none";
animation-name: "initial";
animation-name: "inherit";
animation-name: "unset";
animation-name: "revert";
animation-name: "revert-layer";
```

**Comment by javifernandez** [comment]
> [@javifernandez](https://github.com/javifernandez) So all of these are valid, right?
> 
> ```
> animation-name: "default";
> animation-name: "none";
> animation-name: "initial";
> animation-name: "inherit";
> animation-name: "unset";
> animation-name: "revert";
> animation-name: "revert-layer";
> ```


Yes, that's my understanding now; and what's defined in the current WPT test cases.

**Comment by javifernandez** [comment]
I think all the browser are interoperable now, and no need to change the spec at all, so I believe we can close this issue now.

The only related pending issue would be perhaps #7762, which already has PRs for the spec and the WPT.

**Comment by cdoublev** [comment]
> I think all the browser are interoperable now [...]

This seems true only for `animation-name`. Keyframe names also exist in `@keyframes`'s prelude, and are serialized in `CSSKeyframesRule.cssText` and `CSSKeyframesRule.name`.

Based from my observations in [this CodePen](https://codepen.io/creativewave/full/ByjYYrN):

  - Chrome serializes the `<string>` value as an identifier in `@keyframes`, **even if it is an excluded identifier**, and as is in `CSSKeyframesRule.name`
  - Firefox serializes the `<string>` value as an identifier in `@keyframes`, except if it is an excluded identifier, and as is in `CSSKeyframesRule.name`
  - Chrome and Firefox accept any value for `CSSKeyframesRule.name` and serializes it as specified
  - Firefox replaces `""` with `none` in `animation-name` and `@keyframes`, and serializes it as the empty string in `CSSKeyframesRule.name`

The non-conforming behaviors with `""` and `"none"` are unrelated to this issue.

For what is worth, these are the WPT tests related to this issue:

  1. [`animation-name`](https://github.com/web-platform-tests/wpt/blob/master/css/css-animations/parsing/animation-name-valid.html)
  2. [`@keyframes`](https://github.com/web-platform-tests/wpt/blob/master/css/css-animations/parsing/keyframes-name-valid.html)
  3. [`CSSKeyframesRule.name`](https://github.com/web-platform-tests/wpt/blob/master/css/cssom/CSSKeyframesRule.html)

(1) seems complete.

(2) does not test CSS-wide keywords, `default`, and `none`.

(3) does not test `<string>` with a CSS-wide keyword, `default`, or `none` as its value, or a value that is not an identifier or `<string>`.

**Comment by tabatkins** [comment]
I'm not clear at this point what our resolution is, and what, if anything needs edits (specs and/or tests).

**Comment by javifernandez** [comment]
> I'm not clear at this point what our resolution is, and what, if anything needs edits (specs and/or tests).

My understanding is that the [resolution](https://github.com/w3c/csswg-drafts/issues/2435#issuecomment-428646888) means that the value of the 'animation-name' should be serialized as 'identifies' when possible, and as ```<string>``` otherwise. IMO, there is no need to touch the spec, which [states](https://drafts.csswg.org/css-animations/#at-ruledef-keyframes) as follows:

> When serialized, the value is serialized as an [`<ident>`](https://drafts.csswg.org/css-values-4/#typedef-ident) unless it’s a disallowed keyword, in which case it’s serialized as a `<string>`.

This is how Firefox and Safari interpreted the resolution mentioned before, so they adapted their implementation accordingly. I have recently implemented the same behavior in Chrome. 

However, the last comment mentions some lack of interop in other usages of the ```@keyframe```  value and some missing test cases. As far as I could understand, no spec change is needed.

**Comment by cdoublev** [comment]
> My understanding is that the resolution the [resolution](https://github.com/w3c/csswg-drafts/issues/2435#issuecomment-428646888) means that the value of the 'animation-name' should be serialized as 'identifier' when possible, and as `<string>` otherwise

This resolution is *"serialize **all keyframes** as identifiers"*, not just in `animation-name`. And the requirement to serialize *"as 'identifier' when possible, and as `<string>` otherwise"* is currently specified for `<keyframes-name>`, which is not only used in `animation-name`, but also in other places. 

In these other places, browsers do not always serialize as 'identifier' when possible. So authors could complain to these browsers that their serialization is non-conforming, if the spec does not define how `<keyframes-name>` must serialize in these other places.

I am sorry if I misunderstood this resolution.
